### ISSUE TITLE
unexpected results for predicting on validation sets in python

### ISSUE THREAD

**Issue by ivinogra** [microsoft/LightGBM#3580]
If I understand correctly, Lightgbm makes it difficult to predict on a validation dataset (without access to raw say numpy data). There are functions `Booster.{eval,eval_train,eval_valid}`, but they are not correctly documented and are not user-friendly. 

incorrect doc: By default, `Booster.valid_sets` will be cleared at the end of training making it impossible to predict on training or validation sets using the above methods. The parameter `keep_training_booster` for `train` has to be set to `True` to keep the link. In the documentation for this parameter, there is no word about this effect. I can get around this by defining my own `Booster` and calling `Booster.update`, but this feels like a hack. 

not user-friendly: there is no straightforward way to get the actual predictions - I can `feval` scored on them in a direct way. Again, I can hack an `feval` that changes a global object to get around this or use `Booster._Booster__inner_predict`. 

Please correct me and accept my apologies if any of the statements above are incorrect. Is it a reasonable ask to keep the prediction possibility whenever possible and document releasing datasets accurately?

**Comment by guolinke** [comment]
These interfaces are for the LightGBM itself to use, not for exposing. There are some dependencies, like `__inner_predict_buffer` need to set and reset, and not rival to use them directly. 

If you really when you use them,  the feval function is the best solution.

**Comment by ivinogra** [comment]
> These interfaces are for the LightGBM itself to use, not for exposing. 

Yes, I didn't fully appreciate this point. But I see now that the ways things are setup this is mostly for callbacks. 

> If you really when you use them, the feval function is the best solution.

Not sure I understand your suggestion @guolinke Actual predictions are superior to `feval`: i can plot them, I can share them with another person, I can decide on a suitable metric later, etc. Currently they are released by default without any mention in the documentation, and the usage of methods `Booster.{eval,eval_train,eval_valid}` says nothing about this issue either. Does this deserve to be mentioned somewhere?

Related question: is it possible to produce predictions on a dataset that has correct bins without access to the raw data? [I know the answer is "no" at the moment, unless the predictions are cached, but I think this is theoretically possible and maybe worth implementing?]
